core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala

```diff
@@ -35,16 +37,28 @@ case class ExecutorCacheTaskLocation(override val host: String, executorId: Stri
 
 /**
  * A location on a host.
+ * The host names will be resolved to IP addresses in case SPARK_LOCAL_HOSTNAME is unset
+ * to fix the task locality levels always being 'ANY' issue mentioned in SPARK-10149.
  */
 private [spark] case class HostTaskLocation(override val host: String) extends TaskLocation {
-  override def toString: String = host
+  override def toString: String = if ( TaskLocation.sparkLocalHostname == None ) {
```

Comment:
I think this is important too, I just don't know enough to evaluate the change. BTW these should be something more like `host.getOrElse(InetAddress.getByName(host).getHostAddress)`

I'd like to summon people like ... @andrewor14 @sryza @cmccabe. I do suspect this will be a problem in practice for several use cases.